File: Windows/BotSpeaker/InterruptionMonitor.cs
using System.ComponentModel;
using System.Runtime.CompilerServices;
using System.Windows;
using NAudio.CoreAudioApi;
using NAudio.Wave;

namespace BotSpeaker;

/// <summary>
/// Adaptive interruption detection: learns the monitored input's ambient level,
/// flags activity when a 0.4 s rolling window stays above an adaptive threshold,
/// and clears it after 0.75 s continuously near ambient — same tuning as the
/// macOS InterruptionMonitor.
/// </summary>
public sealed class InterruptionMonitor : INotifyPropertyChanged
{
    public event PropertyChangedEventHandler? PropertyChanged;

    private bool _isMonitoring;
    public bool IsMonitoring { get => _isMonitoring; private set => Set(ref _isMonitoring, value); }

    private bool _isHearingAudio;
    public bool IsHearingAudio { get => _isHearingAudio; private set => Set(ref _isHearingAudio, value); }

    private string? _errorMessage;
    public string? ErrorMessage { get => _errorMessage; private set => Set(ref _errorMessage, value); }

    public Action<bool>? OnActivityChanged { get; set; }

    private WasapiCapture? _capture;
    private double? _monitoringStartedAt;
    private readonly List<float> _calibrationLevels = [];
    private float? _ambientEstimateDb;
    private double? _activityWindowStartedAt;
    private readonly List<float> _activityLevels = [];
    private double? _ambientWindowStartedAt;
    private readonly List<float> _ambientReturnLevels = [];

    private const double CalibrationDuration = 0.75;
    private const double ActivityWindowDuration = 0.4;
    private const double AmbientReturnWindowDuration = 0.75;
    private const float ActivityMarginDb = 10;
    private const float AmbientReturnMarginDb = 4;
    private const float MinimumActivityThresholdDb = -45;
    private const float MaximumActivityThresholdDb = -18;
    private const float AmbientSmoothing = 0.04f;

    private static readonly Guid IeeeFloatSubFormat = new("00000003-0000-0010-8000-00aa00389b71");

    public void Start(string inputId)
    {
        Stop(clearError: false);
        ErrorMessage = null;

        try
        {
            var device = AudioDeviceManager.Device(
                string.IsNullOrEmpty(inputId) ? AudioDeviceManager.DefaultInputDeviceId() ?? "" : inputId)
                ?? throw new AppException("The interruption input is unavailable.");

            var capture = new WasapiCapture(device);
            capture.DataAvailable += OnDataAvailable;
            capture.RecordingStopped += (_, _) => device.Dispose();
            _capture = capture;
            _monitoringStartedAt = Now();
            capture.StartRecording();
            IsMonitoring = true;
        }
        catch (Exception error)
        {
            Stop(clearError: false);
            ErrorMessage = error.Message;
        }
    }

    public void Stop(bool clearError = true)
    {
        if (_capture is not null)
        {
            var capture = _capture;
            _capture = null;
            capture.DataAvailable -= OnDataAvailable;
            try { capture.StopRecording(); } catch (Exception) { }
            capture.Dispose();
        }
        _monitoringStartedAt = null;
        _calibrationLevels.Clear();
        _ambientEstimateDb = null;
        _activityWindowStartedAt = null;
        _activityLevels.Clear();
        _ambientWindowStartedAt = null;
        _ambientReturnLevels.Clear();
        IsMonitoring = false;
        SetHearingAudio(false);
        if (clearError) ErrorMessage = null;
    }

    private void OnDataAvailable(object? sender, WaveInEventArgs args)
    {
        var format = (sender as WasapiCapture)?.WaveFormat;
        if (format is null || args.BytesRecorded == 0) return;

        double sum = 0;
        int sampleCount = 0;

        bool isFloat = format.Encoding == WaveFormatEncoding.IeeeFloat
            || (format is WaveFormatExtensible extensible && extensible.SubFormat == IeeeFloatSubFormat);
        if (isFloat)
        {
            for (int i = 0; i + 4 <= args.BytesRecorded; i += 4)
            {
                float value = BitConverter.ToSingle(args.Buffer, i);
                sum += value * value;
                sampleCount++;
            }
        }
        else if (format.BitsPerSample == 16)
        {
            for (int i = 0; i + 2 <= args.BytesRecorded; i += 2)
            {
                float value = BitConverter.ToInt16(args.Buffer, i) / 32768f;
                sum += value * value;
                sampleCount++;
            }
        }
        else
        {
            return;
        }

        if (sampleCount == 0) return;
        float rms = (float)Math.Sqrt(sum / sampleCount);
        float decibels = 20f * (float)Math.Log10(Math.Max(rms, 0.0000001f));
        Application.Current?.Dispatcher.BeginInvoke(() => Receive(decibels));
    }

    private static double Now() => Environment.TickCount64 / 1000.0;

    private void Receive(float decibels)
    {
        double now = Now();
        if (_monitoringStartedAt is not double startedAt) return;

        if (now - startedAt < CalibrationDuration)
        {
            _calibrationLevels.Add(decibels);
            _ambientEstimateDb = Percentile(_calibrationLevels, 0.3);
            return;
        }

        _ambientEstimateDb ??= decibels;
        float ambient = _ambientEstimateDb.Value;
        float activityThreshold = Math.Min(
            Math.Max(ambient + ActivityMarginDb, MinimumActivityThresholdDb),
            MaximumActivityThresholdDb);

        if (IsHearingAudio)
        {
            CollectAmbientReturn(decibels, now, ambient);
        }
        else
        {
            CollectPossibleActivity(decibels, now, activityThreshold);
            if (_activityWindowStartedAt is null && decibels < activityThreshold)
            {
                _ambientEstimateDb = ambient + (decibels - ambient) * AmbientSmoothing;
            }
        }
    }

    private void CollectPossibleActivity(float decibels, double now, float threshold)
    {
        if (_activityWindowStartedAt is not double startedAt)
        {
            if (decibels >= threshold)
            {
                _activityWindowStartedAt = now;
                _activityLevels.Clear();
                _activityLevels.Add(decibels);
            }
            return;
        }

        _activityLevels.Add(decibels);
        if (now - startedAt < ActivityWindowDuration) return;

        if (AveragePowerDb(_activityLevels) >= threshold)
        {
            _activityWindowStartedAt = null;
            _activityLevels.Clear();
            _ambientWindowStartedAt = null;
            _ambientReturnLevels.Clear();
            SetHearingAudio(true);
        }
        else if (decibels >= threshold)
        {
            _activityWindowStartedAt = now;
            _activityLevels.Clear();
            _activityLevels.Add(decibels);
        }
        else
        {
            _activityWindowStartedAt = null;
            _activityLevels.Clear();
        }
    }

    private void CollectAmbientReturn(float decibels, double now, float ambient)
    {
        float returnThreshold = ambient + AmbientReturnMarginDb;
        if (decibels > returnThreshold)
        {
            _ambientWindowStartedAt = null;
            _ambientReturnLevels.Clear();
            return;
        }

        if (_ambientWindowStartedAt is not double startedAt)
        {
            _ambientWindowStartedAt = now;
            _ambientReturnLevels.Clear();
            _ambientReturnLevels.Add(decibels);
            return;
        }

        _ambientReturnLevels.Add(decibels);
        if (now - startedAt < AmbientReturnWindowDuration) return;
        if (AveragePowerDb(_ambientReturnLevels) > returnThreshold) return;

        float returnedAmbient = AveragePowerDb(_ambientReturnLevels);
        _ambientEstimateDb = ambient + (returnedAmbient - ambient) * 0.2f;
        _ambientWindowStartedAt = null;
        _ambientReturnLevels.Clear();
        _activityWindowStartedAt = null;
        _activityLevels.Clear();
        SetHearingAudio(false);
    }

    private static float AveragePowerDb(List<float> levels)
    {
        if (levels.Count == 0) return -160;
        double meanPower = levels.Average(db => Math.Pow(10, db / 10.0));
        return 10f * (float)Math.Log10(Math.Max(meanPower, 1e-16));
    }

    private static float? Percentile(List<float> levels, double fraction)
    {
        if (levels.Count == 0) return null;
        var sorted = levels.Order().ToList();
        int index = Math.Min((int)((sorted.Count - 1) * fraction), sorted.Count - 1);
        return sorted[index];
    }

    private void SetHearingAudio(bool value)
    {
        if (IsHearingAudio == value) return;
        IsHearingAudio = value;
        OnActivityChanged?.Invoke(value);
    }

    private void Set<T>(ref T field, T value, [CallerMemberName] string? name = null)
    {
        if (EqualityComparer<T>.Default.Equals(field, value)) return;
        field = value;
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(name));
    }
}

File: macOS/BotSpeaker/InterruptionMonitor.swift
import AVFoundation
import AudioToolbox
import Combine
import Foundation

@MainActor
final class InterruptionMonitor: ObservableObject {
    @Published private(set) var isMonitoring = false
    @Published private(set) var isHearingAudio = false
    @Published private(set) var errorMessage: String?

    var onActivityChanged: ((Bool) -> Void)?

    private let engine = AVAudioEngine()
    private var tapInstalled = false
    private var monitoringStartedAt: TimeInterval?
    private var calibrationLevels: [Float] = []
    private var ambientEstimateDB: Float?
    private var activityWindowStartedAt: TimeInterval?
    private var activityLevels: [Float] = []
    private var ambientWindowStartedAt: TimeInterval?
    private var ambientReturnLevels: [Float] = []

    private let calibrationDuration: TimeInterval = 0.75
    private let activityWindowDuration: TimeInterval = 0.4
    private let ambientReturnWindowDuration: TimeInterval = 0.75
    private let activityMarginDB: Float = 10
    private let ambientReturnMarginDB: Float = 4
    private let minimumActivityThresholdDB: Float = -45
    private let maximumActivityThresholdDB: Float = -18
    private let ambientSmoothing: Float = 0.04

    func start(inputUID: String) async {
        stop(clearError: false)
        errorMessage = nil

        let authorized: Bool
        switch AVCaptureDevice.authorizationStatus(for: .audio) {
        case .authorized:
            authorized = true
        case .notDetermined:
            authorized = await AVCaptureDevice.requestAccess(for: .audio)
        default:
            authorized = false
        }

        guard authorized else {
            errorMessage = "Microphone access is required for interruption detection."
            return
        }

        do {
            let inputNode = engine.inputNode
            if !inputUID.isEmpty {
                guard let deviceID = AudioDeviceManager.deviceID(forUID: inputUID),
                      let unit = inputNode.audioUnit else {
                    throw AppError("The interruption input is unavailable.")
                }
                var mutableDeviceID = deviceID
                let status = AudioUnitSetProperty(
                    unit,
                    kAudioOutputUnitProperty_CurrentDevice,
                    kAudioUnitScope_Global,
                    0,
                    &mutableDeviceID,
                    UInt32(MemoryLayout<AudioDeviceID>.size)
                )
                guard status == noErr else {
                    throw AppError("Could not monitor that input (Core Audio \(status)).")
                }
            }

            let format = inputNode.outputFormat(forBus: 0)
            guard format.channelCount > 0, format.sampleRate > 0 else {
                throw AppError("The interruption input has no active audio channels.")
            }

            inputNode.installTap(onBus: 0, bufferSize: 1_024, format: format) { [weak self] buffer, _ in
                guard let self, let channels = buffer.floatChannelData else { return }
                let frameCount = Int(buffer.frameLength)
                let channelCount = Int(buffer.format.channelCount)
                guard frameCount > 0, channelCount > 0 else { return }

                var sum: Float = 0
                for channel in 0..<channelCount {
                    let samples = channels[channel]
                    for frame in 0..<frameCount {
                        let value = samples[frame]
                        sum += value * value
                    }
                }
                let rms = sqrt(sum / Float(frameCount * channelCount))
                let decibels = 20 * log10(max(rms, 0.000_000_1))
                DispatchQueue.main.async { [weak self] in
                    self?.receive(decibels: decibels)
                }
            }
            tapInstalled = true
            monitoringStartedAt = ProcessInfo.processInfo.systemUptime
            engine.prepare()
            try engine.start()
            isMonitoring = true
        } catch {
            stop(clearError: false)
            errorMessage = error.localizedDescription
        }
    }

    func stop(clearError: Bool = true) {
        if tapInstalled {
            engine.inputNode.removeTap(onBus: 0)
            tapInstalled = false
        }
        engine.stop()
        monitoringStartedAt = nil
        calibrationLevels.removeAll()
        ambientEstimateDB = nil
        activityWindowStartedAt = nil
        activityLevels.removeAll()
        ambientWindowStartedAt = nil
        ambientReturnLevels.removeAll()
        isMonitoring = false
        setHearingAudio(false)
        if clearError { errorMessage = nil }
    }

    private func receive(decibels: Float) {
        let now = ProcessInfo.processInfo.systemUptime
        guard let monitoringStartedAt else { return }

        if now - monitoringStartedAt < calibrationDuration {
            calibrationLevels.append(decibels)
            ambientEstimateDB = percentile(calibrationLevels, fraction: 0.3)
            return
        }

        if ambientEstimateDB == nil { ambientEstimateDB = decibels }
        let ambient = ambientEstimateDB ?? decibels
        let activityThreshold = min(
            max(ambient + activityMarginDB, minimumActivityThresholdDB),
            maximumActivityThresholdDB
        )

        if isHearingAudio {
            collectAmbientReturn(decibels: decibels, now: now, ambient: ambient)
        } else {
            collectPossibleActivity(decibels: decibels, now: now, threshold: activityThreshold)
            if activityWindowStartedAt == nil, decibels < activityThreshold {
                ambientEstimateDB = smoothedAmbient(current: ambient, sample: decibels)
            }
        }
    }

    private func collectPossibleActivity(decibels: Float, now: TimeInterval, threshold: Float) {
        guard let startedAt = activityWindowStartedAt else {
            if decibels >= threshold {
                activityWindowStartedAt = now
                activityLevels = [decibels]
            }
            return
        }

        activityLevels.append(decibels)
        guard now - startedAt >= activityWindowDuration else { return }

        if averagePowerDB(activityLevels) >= threshold {
            activityWindowStartedAt = nil
            activityLevels.removeAll()
            ambientWindowStartedAt = nil
            ambientReturnLevels.removeAll()
            setHearingAudio(true)
        } else {
            activityWindowStartedAt = decibels >= threshold ? now : nil
            activityLevels = decibels >= threshold ? [decibels] : []
        }
    }

    private func collectAmbientReturn(decibels: Float, now: TimeInterval, ambient: Float) {
        let returnThreshold = ambient + ambientReturnMarginDB
        guard decibels <= returnThreshold else {
            ambientWindowStartedAt = nil
            ambientReturnLevels.removeAll()
            return
        }

        if ambientWindowStartedAt == nil {
            ambientWindowStartedAt = now
            ambientReturnLevels = [decibels]
            return
        }

        ambientReturnLevels.append(decibels)
        guard let startedAt = ambientWindowStartedAt,
              now - startedAt >= ambientReturnWindowDuration,
              averagePowerDB(ambientReturnLevels) <= returnThreshold else { return }

        let returnedAmbient = averagePowerDB(ambientReturnLevels)
        ambientEstimateDB = ambient + (returnedAmbient - ambient) * 0.2
        ambientWindowStartedAt = nil
        ambientReturnLevels.removeAll()
        activityWindowStartedAt = nil
        activityLevels.removeAll()
        setHearingAudio(false)
    }

    private func smoothedAmbient(current: Float, sample: Float) -> Float {
        current + (sample - current) * ambientSmoothing
    }

    private func averagePowerDB(_ levels: [Float]) -> Float {
        guard !levels.isEmpty else { return -160 }
        let meanPower = levels.reduce(Float.zero) { partial, decibels in
            partial + pow(10, decibels / 10)
        } / Float(levels.count)
        return 10 * log10(max(meanPower, 0.000_000_000_000_000_1))
    }

    private func percentile(_ levels: [Float], fraction: Double) -> Float? {
        guard !levels.isEmpty else { return nil }
        let sorted = levels.sorted()
        let index = min(Int(Double(sorted.count - 1) * fraction), sorted.count - 1)
        return sorted[index]
    }

    private func setHearingAudio(_ value: Bool) {
        guard isHearingAudio != value else { return }
        isHearingAudio = value
        onActivityChanged?(value)
    }
}
